### PR TITLE
[JENKINS-61211] Use userDetailsCache for impersonation

### PR DESCRIPTION
**Pull request by Dohbedoh**
See [JENKINS-61211](https://issues.jenkins-ci.org/browse/JENKINS-61211):

* Propose to use the `UserDetailsCache` for impersonation
* Ability to disable the use of the cache through a system property `hudson.model.User.disableCacheForImpersonation` (also modifiable via Groovy)

One thing that is not clear to me is if there are security implication: Are they some use cases where the impersonation MUST request the security realm ? In which case we could add a new method `getUserDetailsForImpersonation(boolean useCache)` but default the behavior of `getUserDetailsForImpersonation()` to not use the cache.

### Proposed changelog entries

* Use userDetails cache for Impersonation

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees @daniel-beck @rsandell 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Review by res0nance — COMMENTED** [review]
This Looks okay to me but I don't see the value of having a function that allows a cache bypass. But that might be required depending on the answer to the question.

> Are they some use cases where the impersonation MUST request the security realm ?

I'd prefer if there wasn't such a case but personally I'm not sure. It should be easy for users of the API to do the most optimal thing.

File: core/src/main/java/hudson/model/User.java
Comment (Wadeck) [inline core/src/main/java/hudson/model/User.java]:
```suggestion
     * This method checks with {@link UserDetailsCache} and {@link SecurityRealm} if the user is a valid user that 
```

nit: correct JavaDoc

**Review by Wadeck — COMMENTED** [review]
> If {@link SecurityRealm} is a kind that does not support querying information about other users, this will use {@link LastGrantedAuthoritiesProperty} to pick up the granted authorities as of the last time the user has logged in.

This JavaDoc part is wrong. If the cache is enabled, the SecurityRealm will have to load the username otherwise it's an exception. It's only in the case the cache is disabled that the LastGrantedAuthotiries will be used.

Due to the security implication of this, the Security Team needs to have a deeper look on the implication. So I will put the PR on hold as we do not have time yet.

**Comment by jeffret-b** [comment]
I've looked into some odd behaviors related to users and the UserDetailsCache several times before. Caution is warranted. I'll try to take a look at this in the next couple of days.

**Comment by Dohbedoh** [comment]
@jeffret-b Looking through the test failure I have noticed 1 things: the `userDetailsCache` has a cache that record the existence (**and non existence**) of a user. And this apparently does not fit well for this scenario. The non existence results in a `UsernameNotFoundException` that is different from a `UserMayOrMayNotExistException` that is handled later on: https://github.com/jenkinsci/jenkins/blob/jenkins-2.222/core/src/main/java/hudson/model/User.java#L390-L398.

So I think that to make this solution works, I would need to either `bypass` the existence cache or find a way through error handling. I need to spend some time on this.

**Comment by jeffret-b** [comment]
@Dohbedoh, yes those are some of the odd behaviors. There are a number of weird things going on specifically involving the UserDetailsCache and also in related code. I've been back and forth through it many times and it's hard to follow. There are a number of different scenarios that pass through code in this area, including the type of user, the provider, and how the connection is made.

**Comment by jeffret-b** [comment]
@Dohbedoh, do you consider this a work-in-progress at this point? Do you need reviews on its current state?

**Comment by Dohbedoh** [comment]
@jeffret-b yes please set it in WIP.

File: core/src/main/java/jenkins/security/UserDetailsCache.java
Comment (Dohbedoh) [inline core/src/main/java/jenkins/security/UserDetailsCache.java]: When `UserMayOrMayNotExistException` was thrown, we were flagging the user as non existent in the cache. And the next time impersonation was requested for the user, it would result in a different exception [UsernameNotFoundException](https://github.com/jenkinsci/jenkins/blob/stable-2.222/core/src/main/java/jenkins/security/UserDetailsCache.java#L124) that is handled differently by the impersonation [User#getUserDetailsForImpersonation](https://github.com/jenkinsci/jenkins/blob/stable-2.222/core/src/main/java/hudson/model/User.java#L392). 

It seems inconsistent and was the reason for the tests to fail (when using the cache the `UsernameNotFoundException` would make the impersonation fail whereas when no cache is used the `UserMayOrMayNotExistException` is raised and the impersonation does not fail).

I feel like this is a correct fix, it may however impact the efficiency of the cache for user lookup as it may discarding request for both `UsernameNotFoundException` and the subclass `UserMayOrMayNotExistException`. So with this change, the existence cache hit ratio will be reduced. I need a review on this.

Note: The method [UserDetailsCache#getCached](https://github.com/jenkinsci/jenkins/blob/stable-2.222/core/src/main/java/jenkins/security/UserDetailsCache.java#L102) throw the exception `UserMayOrMayNotExistException`. Maybe this is a method that we could use for this improvement.

File: core/src/main/java/jenkins/security/ImpersonatingUserDetailsService.java
Comment (jtnord) [inline core/src/main/java/jenkins/security/ImpersonatingUserDetailsService.java]: unused outside of this class.

```suggestion
    private static boolean DISABLE_CACHE_FOR_IMPERSONATION = 
```

File: core/src/main/java/jenkins/security/ImpersonatingUserDetailsService.java
Comment (Dohbedoh) [inline core/src/main/java/jenkins/security/ImpersonatingUserDetailsService.java]:
```suggestion
            Boolean.getBoolean(ImpersonatingUserDetailsService.class.getName() + ".disableImpersonationCache");
```

**Comment by Dohbedoh** [comment]
@jeffret-b this is not WIP anymore.

**Comment by jeffret-b** [comment]
I removed the wip markers. I'll take a look at when I can.

File: core/src/main/java/jenkins/security/UserDetailsCache.java
Comment (jeffret-b) [reply core/src/main/java/jenkins/security/UserDetailsCache.java]: This change makes sense. It results in the cache containing a more accurate representation of the known state. If the UserMayOrMayNotExist it doesn't make sense to populate the cache with a statement that the user doesn't exist.

It's true that this may change the cache hit ratio. For at least some of the more popular providers with a recommended configuration this is probably an unusual configuration. With providers that routinely throw UserMayOrMayNotExistException this could result in observably different cache behavior. For example, with LDAP without the configuration to search for users.

Still, this results in a more accurate cache, without erroneously declaring that the user does not exist. It's possible that the less accurate cache is sufficient for the existing use, while insufficient for the new one. Changing it could make the existing use slower while not adding data meaningful for that case.

No matter how I look at it I conclude that this proposal is better by more correctly representing the state.

**Review by jeffret-b — APPROVED** [review]
The code looks good. I've played around with it and with some of the existing auto tests. As seen in previous failures there are existing tests that cover things in this area. It's not clear how to add more tests around this change, as it is intended to retain the same behavior but do it faster.

I set up a Jenkins system as described in the Jira ticket. It's easy to reproduce the problem. The ticket describes the method call repeated 8 times. In my case, I observed something much closer to 50. With the HudsonPrivateSecurityRealm I didn't observe significant time savings from the caching. When I simulated a slower, external provider by inserting a sleep in `HudsonPrivateSecurityRealm.loadUserByUsername`, though, then the time with the cache remained the same but took much longer without it.

It would be nice to make changes somewhere else to reduce the number of calls. That doesn't look very straightforward or feasible. It's possible that even with the caching, the number of calls hitting the cache could still cause performance issues. With the cache in place we can continue to evolve if needed.

Ultimately, we need to improve our user caching in several ways. Some of these will probably include extending the SecurityRealm, as mentioned here. We need to do a better job of managing these caches and knowing what we have and its reliability. Concerns about these more complete needs kept me looking for reasons why this PR would be a bad idea, why it might not go far enough. It doesn't solve everything certainly, but it makes a definite improvement in this area. I have not been able to detect any problems this introduces or reasons why it would prevent fixing other issues.

This approach is nicely self-contained for this issue, making significant performance improvements.

**Comment by Wadeck** [comment]
Just a quick review is need after my initial look at it when it was only one-commit long ;)

☁️ 🐝 internal ticket: JENSEC-974

**Comment by Wadeck** [comment]
**Concurrency issue**
What about this execution flow?

- 1) ImpersonatingUserDetailsService#loadUserByUsername
- 2) `UserDetailsCache.get().loadUserByUsername(username)`
- 3) existenceCache ok **but `detailsCache` does not contain the UserDetails**
- 4) calling Retriever => `jenkins.getSecurityRealm().loadUserByUsername(idOrFullName)`
- 5) But for AbstractPasswordBasedSecurityRealm like, the userDetails is delegated to... ImpersonatingUserDetailsService
- 6) Loop to 1) until `existenceCache` and `detailsCache` are sync. 

The condition is that `existenceCache` contains a true but the `detailsCache` does not contain the details. This could happen:
- a) due to the SoftReference used in detailsCache, one could be evicted => temporary issue for up to 2 minutes. I suppose every calls will result in StackOverflowError.
- b) due to a un-sync due to concurrent modification => seems to only cause some additional calls but the loop should not go into SOE

From my PoV, the cache system is not thread safe but as we did not rely on its data consistency but just as a helper it's fine. But now, using it inside a loop could reveal the "thread unsafeness" and causes problems.

@Dohbedoh @jeffret-b Any thoughts on this?

**Comment by jeffret-b** [comment]
@Wadeck, have you verified this flow exists? Or are you suggesting that it might exist?

Also, could provide a little more detail on your step _v_? I'm not sure what you're referring to there as the code seems a little different.

**Comment by Wadeck** [comment]
> have you verified this flow exists? Or are you suggesting that it might exist?

> if this could be addressed (after confirmation)

Not confirmed (and no plan to do it)

---

> Also, could provide a little more detail on your step v? I'm not sure what you're referring to there as the code seems a little different.

> But for AbstractPasswordBasedSecurityRealm like, the userDetails is delegated to... ImpersonatingUserDetailsService

[AbstractPasswordBasedSecurityRealm.java#L42](https://github.com/jenkinsci/jenkins/blob/0e62b68012282648377cfac4fe6d01956d7c1f11/core/src/main/java/hudson/security/AbstractPasswordBasedSecurityRealm.java#L42), as the ImpersonatingUserDetailsService is used by default in such SecurityRealms, relying on the SecurityRealm to provide the value for the cache miss result in sort of recursive calls. The SecurityComponents are the ones responsible to answer method calls like [loadUserByUsername](https://github.com/jenkinsci/jenkins/blob/c059ccb0d834061c302ad642719fcb0764c98a07/core/src/main/java/hudson/security/SecurityRealm.java#L372), at least in default implementation.

Is it more clear like this?

**Comment by jeffret-b** [comment]
Thanks, it does help.

One of my concerns with reusing the UserDetailsCache is that is was created for a specific purpose -- attempting to expand its use may get us into weird situations. On the other hand, I wasn't able to detect any problems when I reviewed. Wadeck's scenario might be possible but I'm not yet convinced it exists.

**Comment by jeffret-b** [comment]
Closing to see if we can get a clean build.

**Comment by Dohbedoh** [comment]
Good point @Wadeck. I understand the scenario and I will dig further to check if I can reproduce this.

**Comment by jeffret-b** [comment]
I'm looking into some other things around this area so I was just reading the Jira ticket attached to this where you write, "Caching can be configured at the Security Realm level depending on the solution (for example for LDAP / Active directory)." Do you have any further information about that? If that's the case, why is this change needed? Does that caching not work sufficiently well? Not apply to all of the realms? Too confusing to implement?

I believe we should have a centralized cache in core for this, but I'm thinking maybe we need something a little more like what exists in the LDAP and Active Directory plugins -- something that has user configuration and visibility. Maybe it would be better to create a new class implementation that replaces the existing one and manages UserDetail caching generally for Jenkins. I'm not sure yet what it would like or how it might be different from this one.

I'm just pondering about the best approach here. WDYT?

**Comment by Dohbedoh** [comment]
@jeffret-b Not all SecurityRealm have a cache implementation. And for those that don't, the `QueueItemAuthenticator` implementations might overload the services behind those Security Realms.

> I believe we should have a centralized cache in core for this, but I'm thinking maybe we need something a little more like what exists in the LDAP and Active Directory plugins -- something that has user configuration and visibility.

I agree. The cache was needed to deal with performance issue introduced by the user lookup. In this ticket, it is about performances when using QueueItemAuthenticators. We could be duplicating effort for each different use cases (like Security Realm). 

@Wadeck made a very good point about that potential loop. If a SecurityRealm extends that `AbstractPasswordBasedSecurityRealm` or does something [similar to it](https://github.com/jenkinsci/jenkins/blob/jenkins-2.238/core/src/main/java/hudson/security/AbstractPasswordBasedSecurityRealm.java#L40-L43) and then use the impersonating service in `loadUserByUsername`:

```
@Override
        public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException, DataAccessException {
            return getSecurityComponents().userDetails.loadUserByUsername(username);
        }
```

It is then easy to fall into that loop. With a simple test case, I was able to reproduce that loop. It is actually failing at the first recursion. The cache implementation detects the recursion overload:

```
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalStateException: Recursive load

	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2234)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3965)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4764)
	at jenkins.security.UserDetailsCache.loadUserByUsername(UserDetailsCache.java:127)
	at jenkins.security.ImpersonatingUserDetailsService.loadUserByUsername(ImpersonatingUserDetailsService.java:45)
	at hudson.model.User.getUserDetailsForImpersonation(User.java:387)
	at hudson.model.User.impersonate(User.java:369)
	at jenkins.security.LastGrantedAuthoritiesPropertyTest.basicFlow(LastGrantedAuthoritiesPropertyTest.java:40)
	[...]
Caused by: java.lang.IllegalStateException: Recursive load
	at com.google.common.base.Preconditions.checkState(Preconditions.java:145)
	at com.google.common.cache.LocalCache$Segment.waitForLoadingValue(LocalCache.java:2330)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2320)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2228)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3965)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4764)
	at jenkins.security.UserDetailsCache.loadUserByUsername(UserDetailsCache.java:127)
	at jenkins.security.ImpersonatingUserDetailsService.loadUserByUsername(ImpersonatingUserDetailsService.java:45)
	at jenkins.security.LastGrantedAuthoritiesPropertyTest$TestSecurityRealm.loadUserByUsername(LastGrantedAuthoritiesPropertyTest.java:99)
	at jenkins.security.UserDetailsCache$Retriever.call(UserDetailsCache.java:172)
	at jenkins.security.UserDetailsCache$Retriever.call(UserDetailsCache.java:161)
	at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4767)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3568)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2350)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2313)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2228)
	... 33 more
```

I have actually noticed that the AD plugin suffered from this when the user lookup was introduced (see [JENKINS-34426](https://issues.jenkins-ci.org/browse/JENKINS-34426?focusedCommentId=257757&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-257757)

My first idea of integrating the case in `User#getUserDetailsForImpersonation` should prevent this problem. It would be more localized to the use case we are trying to solve.

**Comment by jeffret-b** [comment]
@Dohbedoh, is my understanding correct -- you found that the loop Wadeck noticed does exist but it isn't a problem in this case?

If I understood correctly, then my approval still stands.

I still think it could be valuable to address user details caching more holistically, but after spending significant time digging into it I still don't have any good proposals. This approach will work for this issue, at least for now.

**Comment by varyvol** [comment]
@Dohbedoh could you confirm wether @jeffret-b 's understanding about your tests regarding what @Wadeck suggested is correct?

@Wadeck it would be great if you could have a look to Allan's comments as well.

**Comment by Wadeck** [comment]
@varyvol The problem is not solved as last sentence of @Dohbedoh : "My first idea of integrating the case in User#getUserDetailsForImpersonation should prevent this problem. It would be more localized to the use case we are trying to solve."

**Comment by Dohbedoh** [comment]
My fault, I have had the chance to get back to it. I should be able to in the next few weeks.

**Comment by Dohbedoh** [comment]
Actually the loop I was able to reproduce was not correct and not a real world scenario. See https://gist.github.com/Dohbedoh/3f240424dd557b02554d8ebc1442749b. I was doing something like the following in a test and it does not make sense:

```
    private static class LoadUserCountingSecurityRealm extends AbstractPasswordBasedSecurityRealm {
        [...]

        @Override
        public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException, DataAccessException {
            ++counter;
            return this.getSecurityComponents().userDetails.loadUserByUsername(username);
        }

        [...]
    }
```

The test always fails on the first [User#getOrCreateByIdOrFullName](https://gist.github.com/Dohbedoh/3f240424dd557b02554d8ebc1442749b#file-usertest-java-L9). With the changes in this PR, it fails with a `java.lang.IllegalStateException: Recursive load`). Without the changes, it fails with a `StackOverflowError`. Which tells me thatt his scenario is not a real world scenario.

****

The real case that can cause mentioned by @Wadeck mentioned is the case when a record is `true` in `existenceCache` but there is no user details anymore in the `detailsCache`. 

The `UserDetailsCache` uses the `jenkins.getSecurityRealm().loadUserByUsername(idOrFullName)` in case user details cannot be retrieved from the details cache. On some cases where `ImpersonationUserDetailsService` is a security components, that will eventually lead back to the `ImpersonationUserDetailsService`, go through the cache again and loop. I haven't been able to reproduce this behavior with popular Security Realm like the jenkins user database or the LDAP plugin though.

A way to solve this could be to provide a `Supplier` / `Function` down to the `UserDetailsCache.Retriever` so that we can the the cache what to fallback to (could default to `jenkins.getSecurityRealm().loadUserByUsername(idOrFullName)` but in the case of the `ImpersonationUserDetailsService` we would want to supply the `base`).

****

That being said, and in regards to what we have discussed in this thread, I feel that it might be safer to not integrate the cache to the `ImpersonationUserDetailsService` but to the `User#getUserDetailsForImpersonation` instead.

**Review by Wadeck — APPROVED** [review]
This approach is already nicer. With the change of the code location I am not able to find any potential loop ✔️.

FTR it's not possible to have the previous loop in the Embedded security realm due to override of loadUserByUsername, but theoretically possible with LdapSecurityRealm.

----

Just some proposal to make the code cleaner and prevent duplication.

File: core/src/main/java/hudson/model/User.java
Comment (Dohbedoh) [reply core/src/main/java/hudson/model/User.java:431]: I do not expect it to be a public API. Adding the annotation.

File: core/src/main/java/hudson/model/User.java
Comment (Dohbedoh) [reply core/src/main/java/hudson/model/User.java:402]: If we hit `DataAccessException` / `UserMayOrMayNotExistException` when retrieving from the `UserDetailsCache`, we know those come from the `SecurityRealm`. If we call the `ImpersonatingUserDetailsService` in that case, we will have the same result, but that will waste another call to the `SecurityRealm#loadUserByUsername`. Saving that call is the reason for the nested try / catch.
I see you point with the DRY code. I can improve this with your suggestion.

Comment (Wadeck) [reply core/src/main/java/hudson/model/User.java:402]: And with remote calls, that could be annoying in terms of performance. My goal is to prevent two different codes, doing the stuff to be called (exception management in this case), so if you can propose a way that is not calling the security realm twice, it could be better :)

File: core/src/main/java/hudson/model/User.java
Comment (Dohbedoh) [reply core/src/main/java/hudson/model/User.java]: I split the tests and changed the scope to _package_.

**Comment by jglick** [comment]
Revisit after #4848.

**Comment by Dohbedoh** [comment]
@jglick @jeffret-b Rebased from master (after acegi removal). let me know what you think. Should I squash the commits by the way ?

**Comment by jglick** [comment]
Tests fail.

I do not intend to review this.

**Comment by jeffret-b** [comment]
This needs to be squashed when merged. I don't see any reason to do any squashing before that.

I can take a look again after the tests are passing.

**Comment by Dohbedoh** [comment]
Test works locally. Will force push to trigger a build again.

**Comment by jeffret-b** [comment]
> Test works locally. Will force push to trigger a build again.

Or close the PR, wait a minute, and reopen it.

I get the same test failure running it locally.

**Comment by alecharp** [comment]
@Dohbedoh please, don't force-push on PRs which were already reviewed, this creates a situation where new additions to the PR requires any reviewer to look at it as a brand new PR.

Anyway, could you please look at the test failures, they seem to be related the changeset of the PR. Thanks.

**Comment by jglick** [comment]
Also

> force push to trigger a build again

is not necessary to correct flakes. You can either merge new changes from the base branch in, e.g.

```
git pull origin master:master
```

or someone with write permission can trigger a new build using **Re-run** on the `Jenkins` check. In any case, if the test failure looks plausibly related to the PR’s changes, restarting the build is not addressing the problem.

**Comment by alecharp** [comment]
This pull-request had no activity since I added the `needs-fix` label. @Dohbedoh do you have any time to work on this?

**Comment by alecharp** [comment]
Addind the `proposed-for-close` label for now. We should close this PR is the next days. @Dohbedoh if you see that after, we can reopen it if needed.

**Comment by basil** [comment]
I read through the long discussion in this PR several times but I must admit that I still cannot understand whether this PR is complete or whether there were any remaining action items. Since there has not been a response in a long time and since there are merge conflicts and test failures, I am closing this PR for now. Feel free to reopen it to pick up where we left off.

**Pull request closed without merging**
